cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -674,6 +684,53 @@ Status CopyFiles(const std::shared_ptr<FileSystem>& source_fs,
   return CopyFiles(sources, destinations, io_context, chunk_size, use_threads);
 }
 
+struct Registry {
+  std::shared_mutex mutex;
+  std::unordered_map<std::string, FileSystem::Factory*> scheme_to_factory;
+  std::vector<void (*)()> finalizers;
+  bool finalized = false;
+};
+
+extern "C" ARROW_EXPORT auto* GetFileSystemFactoryRegistry() {
+  // Check if this function is the one linked to main
+  static Registry registry;
+  return &registry;
+}
+
+extern "C" {
+ARROW_EXPORT void MergeFileSystemRegistry(void* main_registry) {
+  if (GetFileSystemFactoryRegistry() == main_registry) return;
+
+  auto& [mutex, scheme_to_factory, finalizers, _f] = *GetFileSystemFactoryRegistry();
+  auto& [main_mutex, main_scheme_to_factory, main_finalizers, _mf] =
+      *static_cast<Registry*>(main_registry);
+
+  std::unique_lock lock{mutex}, main_lock{main_mutex};
+
+  for (auto& [scheme, factory] : scheme_to_factory) {
+    auto& ref = main_scheme_to_factory[scheme];
+    if (ref) continue;
+    ref = factory;
+  }
+  scheme_to_factory.clear();
+
+  for (auto* finalizer : finalizers) {
+    main_finalizers.push_back(finalizer);
+  }
+  finalizers.clear();
+}
```

Comment:
This can become `GetFilesystemFactoryRegistry()->Merge(main_registry);` with `Merge` being defined at `Registry`.  The initial check would be `this == main_registry` and all the member variables would be already in scope.